### PR TITLE
soc: arm: Add support for kinetis k2x

### PR DESCRIPTION
**Pull request by ostaebler**
This adds support for Kinetis K22 family, specifically for the
MK22FX512.

This is based on [1] and the PR [2] respectively.

[1] https://github.com/sknsean/zephyr/commits/k22
[2] https://github.com/zephyrproject-rtos/zephyr/pull/11852

Signed-off-by: Oliver Stäbler <oliver.staebler@bytesatwork.ch>

### CODE REVIEW

**Comment by ostaebler** [comment]
Please see the corresponding PR for module/hal

https://github.com/zephyrproject-rtos/hal_nxp/pull/7

File: soc/arm/nxp_kinetis/k2x/Kconfig.soc [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+# Copyright (c) 2019 bytes at work AG
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+choice
+prompt "Kinetis K2x MCU Selection"
+depends on SOC_SERIES_KINETIS_K2X
+
+config SOC_MK22FA12
```
Comment (bfrog) [inline soc/arm/nxp_kinetis/k2x/Kconfig.soc]: Does this need SMC to get HSRUN mode? I ended up having to add HAS_MCUX_SMC here and in the hal for myself to get the full speed clock

File: soc/arm/nxp_kinetis/k2x/soc.c [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#include <init.h>
+
+#if defined(CONFIG_SOC_MK22FA12)
+#include <system_MK22FA12.h>
+#else
+#error "Unknown Soc"
+#endif
+
+static int system_wrapper(struct device *arg)
```
Comment (bfrog) [inline soc/arm/nxp_kinetis/k2x/soc.c]: If you look at the K6X soc.c, the clocks are setup based on Kconfig options here. You have DTS clock divider settings for the SOC above, probably doable with those defines

File: soc/arm/nxp_kinetis/k2x/linker.ld [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ * @file
+ * @brief Linker command/script file
+ *
+ * This is the linker script for both standard images and XIP images.
+ */
+
+#include <autoconf.h>
+
+/*
+ * K64F Flash configuration fields
```
Comment (bfrog) [inline soc/arm/nxp_kinetis/k2x/linker.ld]: should be K2X here

File: soc/arm/nxp_kinetis/k2x/soc.c [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+{
+	ARG_UNUSED(arg);
+
+	BOARD_InitBootClocks();
```
Comment (bfrog) [inline soc/arm/nxp_kinetis/k2x/soc.c]: Where is this implemented?

**Review by teburd — CHANGES_REQUESTED** [review]
I'd like to see the clock initialization done closer to the way k6x soc.c does its clock setup. In fact it looks like its possible to do the clock setup entirely with the DTS defines rather than the Kconfig ones k6x's soc.c uses which I feel like would be cleaner as you can then, per board, overlay your own clock settings as needed.

File: soc/arm/nxp_kinetis/k2x/Kconfig.soc (hunk 1 above)
Comment (ostaebler) [reply soc/arm/nxp_kinetis/k2x/Kconfig.soc]: The SOC I use (fx512) does not support HSRUN, so I dropped the SMC support. But for other MK22 I think you need it.

File: soc/arm/nxp_kinetis/k2x/soc.c (hunk 4 above)
Comment (ostaebler) [reply soc/arm/nxp_kinetis/k2x/soc.c]: In `clock_config.c` from SDK.

File: soc/arm/nxp_kinetis/k2x/linker.ld (hunk 3 above)
Comment (ostaebler) [reply soc/arm/nxp_kinetis/k2x/linker.ld]: Fixed

File: soc/arm/nxp_kinetis/k2x/Kconfig.soc (hunk 1 above)
Comment (bfrog) [reply soc/arm/nxp_kinetis/k2x/Kconfig.soc]: Good to know, I'll add that when I add MK22FN512

File: soc/arm/nxp_kinetis/k2x/soc.c (hunk 4 above)
Comment (MaureenHelm) [reply soc/arm/nxp_kinetis/k2x/soc.c]: Please follow the Zephyr convention in other Kinetos SoCs

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Please add a board for this SoC in the same PR; otherwise this code doesn't get built in CI.

**Comment by galak** [comment]
@ostaebler any plans to update this PR?

**Comment by ostaebler** [comment]
> @ostaebler any plans to update this PR?

Hi @galak 
Sorry for my late reply.

Unfortunately, I'll not find the time to update this PR.

**Pull request closed without merging**